### PR TITLE
Generate .dir-locals.el at configuration time.

### PR DESCRIPTION
**Pull request by SkySkimmer**
This is just to auto-select the right path for coqtop.

EDIT: significant changes since the PR opened, see https://github.com/coq/coq/pull/1136#issuecomment-339040603

### CODE REVIEW

**Review by JasonGross — DISMISSED** [review]

**Comment by Zimmi48** [comment]
Thanks! Though, I'll take this opportunity to mention that @letouzey privately mentioned to me that all those `.dir-locals.el` were quite alarming: they send security alerts to the user each time you open a file, until you add them to your `.emacs`. And theoretically they can be used to do very damageable things, although here it's just to set a few Emacs variables. I wonder if there would be a better solution (that would also work cross-editors possibly) using `_CoqProject`.

In any case, this comment should not be interpreted as blocking this PR in any way.

**Comment by SkySkimmer** [comment]
It's possible in `_CoqProject` to put `COQTOP = bla` (coq_makefile only uses COQC so it would be ignored AFAIK), but it doesn't seem to be used by proof general. I'm not sure if it would work well with relative paths as well. Of course arbitrary setting of `coq-prog-name` for people without `proof-prog-name-ask` can do arbitrary things as well.

**Comment by Zimmi48** [comment]
> It's possible in _CoqProject to put COQTOP = bla (coq_makefile only uses COQC so it would be ignored AFAIK), but it doesn't seem to be used by proof general.

Maybe we should open a feature-request for Proof-General?

> Of course arbitrary setting of coq-prog-name for people without proof-prog-name-ask can do arbitrary things as well.

Right, that's something that I didn't even consider.

**Comment by SkySkimmer** [comment]
Actually shouldn't this, the plugins/ and the theories/ dir-locals be merged in the root one? Especially since it means the root one isn't loaded for files in those directories.

**Comment by Zimmi48** [comment]
> Actually shouldn't this, the plugins/ and the theories/ dir-locals be merged in the root one? Especially since it means the root one isn't loaded for files in those directories.

Probably yes. @JasonGross WDYT?

**Comment by JasonGross** [comment]
> Actually shouldn't this, the plugins/ and the theories/ dir-locals be merged in the root one? Especially since it means the root one isn't loaded for files in those directories.

Yes, but that one is a bit more subtle to review.  I'd say merge this one (which seems immediately useful and easy to review), and open a new PR to merge the files.  Subtleties include the fact that plugins and theories bind different directories to the `Coq` name, both pass `-coqlib` (and test-suite doesn't), and both make use of `expand-file-name`, and paths need to be adjusted if the location of `.dir-locals.el` changes.  I suspect the `-R` paths can both be passed, and that passing `-coqlib` in test-suite won't hurt, but that definitely needs testing, which this PR requires less of.

**Comment by maximedenes** [comment]
> Actually shouldn't this, the plugins/ and the theories/ dir-locals be merged in the root one?

@SkySkimmer should I merge this anyway?

**Comment by SkySkimmer** [comment]
Sure, why not.

**Comment by letouzey** [comment]
As @Zimmi48 mentioned above, I object deeply to the use and abuse of `.dir-locals.el` hacks in Coq. Please take the time to consider other solutions. I already noticed PR #386 too late to complain before it got merged, and now I have to cope with it. These `.dir-locals.el` are particularly intrusive for other developers, leading to nasty questions at each file opening unless you cripple your `.emacs` with cryptic code. Moreover these `.emacs` intrusions have frequently to be adapted : already 3 versions of the Coq "root" `.dir-locals.el`, and now two more for theories and plugins, this one about test-suite, what next ? For anyone properly managing its `.emacs` (through git in my case), this is just a pain. Moreover (even if that doesn't frighten me too much here), imposing emacs settings to other users could clearly be a important security risk.

So please don't merge that yet. I'll try to provide decent alternatives. Probably a snippet to include in its .emacs if one choose to, or some better settings directly inside Coq ?

**Review by letouzey — DISMISSED** [review]
As said in my previous comment, I object to the use of `.dir-local.el` for this kind of tweaks.

**Comment by ejgallego** [comment]
Just to add to the discussion as an Emacs user, I do find complex `.dir-locals` quite valuable, even if not perfect, they are the lesser evil to me.

**Comment by SkySkimmer** [comment]
https://github.com/SkySkimmer/proofgeneral/commit/2fbbd5eb1b5d1d6c2f1bc8e1bf2dc117a7f7c164 reads COQTOP.
To replace the theories and plugins dir-locals by _CoqProject PG would also need to understand -coqlib (pretty easy to do).

Does this seem like a good solution?

**Comment by SkySkimmer** [comment]
New dir-locals system with generation of a single toplevel .dir-locals.el at configure time (previous version of this PR at https://github.com/SkySkimmer/coq/tree/test-dir-locals-old https://github.com/SkySkimmer/coq/commit/c2b89f103d5273c45f86f02d5c1c260dc4f19ee1). Generation can be skipped with `-dir-locals no`.

Generation at configure time means we can take advantage of the emacs safe variable system to reduce warnings. We do set some unsafe variables though. See commit message and code for details.

Since this significantly changes the PR I'm going to dismiss-and-request reviews (and change the name of the PR).

**Comment by Zimmi48** [comment]
Not an expert but it looks good to me. It means that I'm going to have to accept Emacs variables for each of my worktrees. I'll do it for the main ones (that track master and v8.7) and not for the temporary ones for which I'll just approve the variables each time I launch Emacs.

**Comment by SkySkimmer** [comment]
>Not an expert but it looks good to me. It means that I'm going to have to accept Emacs variables for each of my worktrees. I'll do it for the main ones (that track master and v8.7) and not for the temporary ones for which I'll just approve the variables each time I launch Emacs.

That sounds pretty painful actually, it's not every time you launch emacs but every time you open a buffer. Or you can do `!` but then it writes in your custom file so if you don't want it permanently you have to go revert it.

**Comment by SkySkimmer** [comment]
I tried unsetting `compilation-search-path` and it seems to find files fine without it. That would leave `camldebug-command-name` as the only unconditionally unsafe value (indeed it is used without prompting).

OTOH `camldebug-command-name` is used in the `camldebug` function only AFAICT https://github.com/ocaml/ocaml/blob/trunk/emacs/camldebug.el#L501. This is relevant because tuareg overrides this function with `ocamldebug` https://github.com/ocaml/tuareg/blob/master/ocamldebug.el#L525 which uses `ocamldebug-command-name` always with prompting.
So I think we should be setting `ocamldebug-command-name` in dir-locals while leaving non-tuareg people alone (and I'll make a PR to tuareg so it's considered safe).

That would leave only conditionally unsafe `compile-command` (safe by default) and `coq-prog-name` (unsafe by default).

File: _CoqProject
Comment (JasonGross) [inline _CoqProject]: I think this is wrong.  I think some of the files rely on `Coq.Init.Prelude` being loaded by default, and rely on `-I plugins` (which I think follows from `-coqlib .`).  Does PG parse `-coqlib` in `_CoqProject`?

File: configure.ml
Comment (JasonGross) [inline configure.ml]: Do you mean `./configure` rather than `../configure`?

**Review by JasonGross — CHANGES_REQUESTED** [review]

File: configure.ml
Comment (SkySkimmer) [reply configure.ml]: >Do you mean ./configure rather than ../configure?

oops

File: _CoqProject
Comment (SkySkimmer) [reply _CoqProject]: In fact `-boot` implies the right `-coqlib`, see https://github.com/coq/coq/blob/master/lib/envars.ml#L121-L124.
The other uses are camlp4(5) detection also in envars.ml, adding stm/ and ide/ to loadpath in https://github.com/coq/coq/blob/master/toplevel/coqinit.ml#L89-L92 and whatever https://github.com/coq/coq/blob/master/tools/coqmktop.ml#L276 does.

I tried a few random files and they worked.

PG doesn't parse coqlib.

Comment (JasonGross) [reply _CoqProject]: Fascinating.

https://github.com/coq/coq/blob/master/tools/coqmktop.ml#L276 is responsible for adding the correct directories to the OCaml loadpath via `-I` when running `coqtop` from an installed location rather than from a local location.  I believe `-boot` has the same effect as `-coqlib .` in this case.

Comment (SkySkimmer) [reply _CoqProject]: @JasonGross do you still want it changed or should I dismiss review?

Comment (JasonGross) [reply _CoqProject]: I'm fine with leaving this.  I think the double dot on configure should be changed, and I plan to change my review to accept once that's updated.  (Though I'll be basically off the grid for a week, so feel free to dismiss my review if you fix that.)

File: configure.ml
Comment (psteckler) [inline configure.ml]: DEVELOPERS

File: configure.ml
Comment (SkySkimmer) [reply configure.ml]: >Do you mean ./configure rather than ../configure?

Ended up using just `configure`.

**Review by JasonGross — APPROVED** [review]

**Comment by cpitclaudel** [comment]
> imposing emacs settings to other users could clearly be a important security risk.

Maybe? But reading form _CoqProject is definitely not safer.

> As @Zimmi48 mentioned above, I object deeply to the use and abuse of .dir-locals.el hacks in Coq. Please take the time to consider other solutions.

@letouzey We're happy to help on the PG side, but do you have concrete solutions to suggest?

Here's one proposal, based on what we do for F*: we could change coq-prog-name to accept a function returning a string as well as a plain string, and have two safe-by-default values: `"coqtop"` and `#'coq-locate-dev-coqtop`.  That function would ship with PG and would do the work of locating coqtop based on the current path.

The Coq repo would then just ship a .dir-locals.el that sets `coq-prog-name` to `#'coq-locate-dev-coqtop`.

**Comment by Matafou** [comment]
Hi, my 2 cents:

@cpitclaudel allowing #'coq-locate-dev-coqtop by default suggest a malicious coq library containing a malicious ./bin/coqtop. Security should precisely ask the user about that just downloaded development that wants to launch its own version of coq.

More generally IMHO using .dir-locals is ok but a good _CoqProject setting is much better. It is IDE-agnostic and we can still implement security checks for that.

So I think we should consider the use of .dir-locals as temporary here and think about something cleaner and safe. Maybe we can ask that the allowed binaries for coqtop are described in .coqrc? And if we support wildcards in .coqrc we could have a default list of binaries containing something like ["coqtop" "~/.opam/*/coqtop"] for instance.

**Comment by cpitclaudel** [comment]
> and we can still implement security checks for that.

Which ones? I'm a bit confused by what we're trying to achieve here. On the one hand we want safety, which requires prompting. On the other hand we don't want to prompt, because @letouzey doesn't like being prompted. What can we do?

**Comment by Matafou** [comment]
You are right we need to satisfy two contradictory goals, that is the curse of all security stuff I am afraid: usability and security are difficult to reconcile.

.dir-locals security works purely syntactically: if you change one character in the .dir-locals it must be prompted again to be allowed, this is a bit extreme it is justified: .dir-locals can contain arbitrary lisp code.

In our case the security-critical information is only a set of binary paths that we accept to lauch from pg. It is possible to make a less obtrusive security mechanism.

In @letouzey case I guess if Pierre accepts to have his clones of coq in a particular directory, say ~/dev/coq it would be nice to put "~/dev/coq/*/coqtop" somewhere and be rid of prompts forever.

**Comment by cpitclaudel** [comment]
> In @letouzey case I guess if Pierre accepts to have his clones of coq in a particular directory, say `~/dev/coq` it would be nice to put `"~/dev/coq/*/coqtop"` somewhere and be rid of prompts forever.

That's already possible, with no changes to PG nor the Coq repo:

```
(put 'coq-prog-name 'safe-local-variable
     (lambda (path) (or (equal path "coqtop")
                   (string-prefix-p "/home/letouzey/dev/coq/"
                                    (expand-file-name path)))))
```

But I imagine that was what @letouzey was referring to when he mentioned crippling his .emacs with cryptic code? Or maybe it's the custom-set-variables form introduced when you mark a value as safe that Pierre doesn't like?

> In our case the security-critical information is only a set of binary paths that we accept to lauch from pg. It is possible to make a less obtrusive security mechanism.

Maybe ? :) But I'm not too sure: as far as I understand the proposal, that's more or less what dir-locals does, isn't it? It prompts you about allowing a particular binary path and if you decide to mark it as safe it stores it in a list.  I think @letouzey finds that list too intrusive. So I'm not sure how we could do much better…

**Comment by cpitclaudel** [comment]
If it helps Coq devs, though, we could certainly make the current predicate more flexible (something like `coq-trusted-paths` and mark any descendent of that path as safe.

@letouzey, would that help?

**Comment by SkySkimmer** [comment]
>But I imagine that was what @letouzey was referring to when he mentioned crippling his .emacs with cryptic code?

Note that at the time of his comment the PR (and the previous state) still used `eval` in dir-locals so the safe value was significantly more complex than a list of strings. This also means that your `put` would not have silenced the warnings.

>@cpitclaudel allowing #'coq-locate-dev-coqtop by default suggest a malicious coq library containing a malicious ./bin/coqtop.

The `eval` form already allows that if the target has already set it safe while using a legitimate Coq development.

**Comment by Matafou** [comment]
> @cpitclaudel allowing #'coq-locate-dev-coqtop by default suggest a malicious coq library containing a malicious ./bin/coqtop.
> The eval form already allows that if the target has already set it safe while using a legitimate Coq development.

Sure, I was pointing the fact that this solution was not really better than before.

**Comment by Matafou** [comment]
> That's already possible, with no changes to PG nor the Coq repo:
> 
> (put 'coq-prog-name 'safe-local-variable
>      (lambda (path) (or (equal path "coqtop")
>                    (string-prefix-p "/home/letouzey/dev/coq/"
>                                     (expand-file-name path)))))

Nice! It is easy to reduce the part of this that should go in .emacs by moving this to PG:
```lisp
(defvar coq-user-safe-path ...)

(put 'coq-prog-name 'safe-local-variable
      (lambda (path) (or (equal path "coqtop")
                    (string-prefix-p coq-user-safe-path
                                     (expand-file-name path)))))
```
and only asking to set ``coq-user-safe-path`` in .emacs.

**Comment by cpitclaudel** [comment]
@Matafou, yup, that was exactly my proposal with `coq-trusted-path` :) Let's wait for an opinion from @letouzey ?

**Comment by SkySkimmer** [comment]
BTW HoTT is using the same approach as Coq for dir-locals (ie an eval form), see eg https://github.com/HoTT/HoTT/blob/master/theories/.dir-locals.el

**Comment by SkySkimmer** [comment]
@letouzey could you give your thoughts on the approach of configure generating the dir-locals as explained in https://github.com/coq/coq/pull/1136#issuecomment-339040603 and the commit message of 173d4f5?

**Comment by SkySkimmer** [comment]
How to progress?

**Comment by Zimmi48** [comment]
It's a shame that discussion about this with @letouzey could not happen at the WG.
I will share the concern that he expressed privately to me a while ago about the current state of this PR.
He is afraid that by making the `.dir-local.el` an artificat of `./configure`, it is going to make it even less stable and thus he will be prompted even more often to accept "unsafe variables".

Do you think that this concern is relevant? If it is not, then it is probably safe to go ahead and merge this.
(At least there will now be the possibility to unable `.dir-local.el` generation.)

I'm personally favorable to this PR as it will also help with #6137.

Finally, we can always discuss this issue again (having to accept new unsafe variables too often), if it happens to be the case.

**Comment by SkySkimmer** [comment]
>Do you think that this concern is relevant?

I think it depends on workflow, ie if one makes temporary Coq directories to work on branches as opposed to just `git checkout` the variable value would change for every directory where the `eval` wouldn't. I have no idea if that's how @letouzey or anyone else works.

OTOH using configure makes it possible to avoid the big eval, which is necessary for @Matafou's code or for putting variables in `ignored-local-variables`, as well as making the overall disable flag possible.

**Comment by Zimmi48** [comment]
That's sometimes how I work but in this case, I will know why it asks me again about these values and will accept them without making them permanent (because I'm also git-tracking my `.emacs`).

**Comment by SkySkimmer** [comment]
One issue I encountered is that if you try to use an installed Coq the -boot in _CoqProject will screw it up.
ie
- make file foo.v at toplevel of the repository
- try to run it with opam Coq or nix Coq or whatever non local Coq you have (eg to see differences between how it and master work)
- witness failure

**Comment by Zimmi48** [comment]
Indeed, with `coqide foo.v`:

```
Warning:
Cannot open /nix/store/z0qfka9n4fdbd1gak7s26g4i80pqcbf5-coq-8.7.0/theories
[cannot-open-path,filesystem]
Warning:
Cannot open /nix/store/z0qfka9n4fdbd1gak7s26g4i80pqcbf5-coq-8.7.0/plugins
[cannot-open-path,filesystem]
Error during initialization: 
Error:
File not found on loadpath : coqidetop.cmxs
Loadpath: /home/theo/coq-pr-1136:/nix/store/z0qfka9n4fdbd1gak7s26g4i80pqcbf5-coq-8.7.0/ide:/nix/store/z0qfka9n4fdbd1gak7s26g4i80pqcbf5-coq-8.7.0/stm
```

Is this a bug?

**Comment by psteckler** [comment]
I've been experiencing this issue with CoqIDE for the past week or so.

I might be wrong, but I think it started with the changes that removed
coqmktop.

**Comment by ejgallego** [comment]
> Is this a bug?

Looks strange, are the files there? Note that in a non-local install the `idetops` won't be installed in `ide` but in `toplevels` or some directory like that, so the loadpath somehow looks wrong for a global install indeed.

[I submitted a bug about this discrepancy sometime ago]

**Comment by maximedenes** [comment]
Ok, so is this ready for merge? The discussion is hard to follow.

**Comment by SkySkimmer** [comment]
@psteckler 
>I've been experiencing this issue with CoqIDE for the past week or so.

Are you somehow using the _CoqProject from this PR? Otherwise please report as a separate issue.

@Zimmi48 
>Is this a bug?

No idea.

@maximedenes 
>Ok, so is this ready for merge? The discussion is hard to follow.

I don't think so.

File: configure.ml
Comment (psteckler) [inline configure.ml]: literal misspelled here

**Comment by SkySkimmer** [comment]
It looks like we're going with #6556 instead.

**Pull request closed without merging**
